// srt2fcpxml/build/main/parser.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Parser = void 0;
const chalk_1 = __importDefault(require("chalk"));
const fs = __importStar(require("fs-extra"));
const subtitle_1 = require("subtitle");
// import { parse, resync, stringify } from 'subtitle';
const uuid_1 = require("uuid");
/* cSpell:disable */
const xml_formatter_1 = __importDefault(require("xml-formatter"));
/* cSpell:enable */
const resources_1 = require("./constants/resources");
const template_1 = require("./template");
function runParse(params) {
    const list = [];
    fs.createReadStream(params.srtPath)
        .pipe(subtitle_1.parse())
        .pipe(subtitle_1.map((node) => {
        if (node.type === 'cue') {
            // eslint-disable-next-line functional/immutable-data
            node.data.text = node.data.text.replace(/"/g, '&#34;');
        }
        return node;
    }))
        .on('data', (node) => {
        // eslint-disable-next-line functional/immutable-data
        list.push(node);
    })
        .on('error', (err) => {
        console.log(chalk_1.default.blue.bgRed.bold(err));
    })
        .on('finish', () => {
        const lastCue = list.slice(-1)[0];
        const totalCueTime = lastCue.data.end;
        const fileName = params.srtPath.replace(/(.+)\/(.+)$/, '$2');
        const projectName = params.projectName || fileName;
        /* cSpell:disable */
        const outputPath = params.outputPath + '/' + fileName + '.fcpxml';
        const eventName = params.eventName || 'srt2fcpxml_node';
        /* cSpell:enable */
        const resourceConfig = resources_1.resources[params.rateKey];
        const gap = params.gap;
        const isUsePlugin = params.isUsePlugin;
        const config = {
            cuesTemplate: template_1.CueTemplate(list, resourceConfig, gap, isUsePlugin).join(''),
            project: {
                name: projectName,
                uid: uuid_1.v4(),
                modDate: 'today',
            },
            event: {
                name: eventName,
                uid: uuid_1.v4(),
            },
            totalCueTime: totalCueTime,
            resources: resourceConfig,
            gap,
            isUsePlugin,
        };
        const fcpXMl = template_1.MainTemplate(config, resourceConfig);
        const fcpXmlFile = fs.createWriteStream(outputPath);
        /* cSpell:disable */
        const formattedXml = xml_formatter_1.default(fcpXMl, {
            indentation: '  ',
            filter: (node) => {
                return node;
            },
            collapseContent: true,
            lineSeparator: '\n',
        });
        fcpXmlFile.write(formattedXml);
        fcpXmlFile.end(() => {
            console.log(chalk_1.default.greenBright.bold(`success! the output file are ${outputPath}`));
        });
    });
}
exports.Parser = (params) => {
    return runParse(params);
};
