// srt2fcpxml/build/main/lib/helper.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.copyPlugin = exports.formateRateKey = exports.checkRate = exports.checkSrtPath = void 0;
/**
 * checkout file exist
 */
const os_1 = __importDefault(require("os"));
const path = __importStar(require("path"));
const chalk_1 = __importDefault(require("chalk"));
const copydir = __importStar(require("copy-dir"));
const fs = __importStar(require("fs-extra"));
const resources_1 = require("../constants/resources");
const log = console.log;
exports.checkSrtPath = (srtPath) => {
    if (!fs.existsSync(srtPath)) {
        log(chalk_1.default.red("the srt file doesn't exist at that path"));
        return false;
    }
    return true;
};
exports.checkRate = (rate) => {
    const keys = Object.keys(resources_1.resourcesEnum);
    if (!keys.includes(rate)) {
        log(chalk_1.default.red("the input rate doesn't match, example: 23.98,24,25,29.87,30,50,59.94,60"));
        return false;
    }
    return true;
};
exports.formateRateKey = (rate) => {
    if (rate) {
        return 'p' + (parseFloat(rate) * 100).toString();
    }
    else {
        return resources_1.resourcesEnum.p3000;
    }
};
exports.copyPlugin = () => {
    const titlesPath = `${os_1.default.homedir()}/Movies/Motion Templates.localized/Titles.localized`;
    const pluginPath = path.resolve(__dirname, '../../../plugin/adaptive-background');
    if (!fs.existsSync(titlesPath)) {
        /* cSpell:disable */
        log(chalk_1.default.red('hey, your FCPX Motion is not a default path, please copy by yourself'));
        /* cSpell:enable */
    }
    copydir.sync(pluginPath, titlesPath + '/adaptive-background', {
        utimes: true,
        mode: true,
        cover: true,
    }, (error) => {
        if (error)
            log(chalk_1.default.red('hey, somethings wrong,please copy plugin by yourself'));
    });
    log(chalk_1.default.greenBright('copy custom plugin done!'));
    return null;
};
